### PR TITLE
CORE-20027 - Added a backoff mechanism to reduce the number of db calls when there are insufficient tokens

### PR DESCRIPTION
**Pull request by filipesoliveira**
The `Token Selection` will always go to the database when there is not enough tokens. This can cause performance degradation. To avoid going to the database too often when there is not enough tokens available, a backoff mechanism has been implemented.

Depends on: https://github.com/corda/corda-api/pull/1595/files

### CODE REVIEW

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt [hunk 1]
```diff
@@ -76,13 +78,17 @@ class TokenCacheEventProcessorFactoryImpl(
             tokenSelectionMetrics
         )
 
+        val dbTokensFetchMinIntervalInMillis = 500L
+        val dbTokensFetchMaxIntervalInMillis = 6000L
```
Comment (vkolomeyko) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt]:
```suggestion
        val dbTokensFetchMaxIntervalInMillis = 60000L
```

Comment (vkolomeyko) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt]: I would imagine these will be read from config.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/handlers/TokenClaimQueryEventHandler.kt [hunk 2]
```diff
@@ -9,20 +9,23 @@ import net.corda.ledger.utxo.token.cache.factories.RecordFactory
...
     private val recordFactory: RecordFactory,
     private val availableTokenService: AvailableTokenService,
-    private val serviceConfiguration: ServiceConfiguration
+    private val serviceConfiguration: ServiceConfiguration,
+    private val backoffManager: BackoffManager
 ) : TokenEventHandler<ClaimQuery> {
 
     private companion object {
-        private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
+        val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
```
Comment (vkolomeyko) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/handlers/TokenClaimQueryEventHandler.kt:27]: Why remove `private`, please?

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt [hunk 3]
```diff
@@ -24,8 +24,10 @@ import net.corda.ledger.utxo.token.cache.services.TokenSelectionMetrics
 import net.corda.ledger.utxo.token.cache.services.TokenSelectionMetricsImpl
 import net.corda.ledger.utxo.token.cache.services.TokenSelectionSyncRPCProcessor
 import net.corda.ledger.utxo.token.cache.services.internal.AvailableTokenServiceImpl
+import net.corda.ledger.utxo.token.cache.services.internal.BackoffManager
 import net.corda.libs.statemanager.api.StateManager
 import net.corda.orm.JpaEntitiesRegistry
+import net.corda.utilities.retry.Exponential
```
Comment (vkolomeyko) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt]: I wonder if we need this `import` here?

**Review by vkolomeyko — COMMENTED** [review]
Looks broadly fine, few minor comments posted.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/handlers/TokenClaimQueryEventHandler.kt (hunk 2 above)
Comment (filipesoliveira) [reply components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/handlers/TokenClaimQueryEventHandler.kt:27]: The companion object is already private.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt (hunk 1 above)
Comment (filipesoliveira) [reply components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt]: Yes, that is the bit left out. I am working on that at the moment.

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt (hunk 3 above)
Comment (filipesoliveira) [reply components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/factories/TokenCacheEventProcessorFactoryImpl.kt]: Old change. It has been removed.

File: gradle.properties [hunk 4]
```diff
@@ -39,7 +39,7 @@ commonsLangVersion = 3.12.0
 commonsTextVersion = 1.10.0
 # Corda API libs revision (change in 4th digit indicates a breaking change)
 # Change to 5.3.0.xx-SNAPSHOT to pick up maven local published copy
-cordaApiVersion=5.3.0.8-beta+
+cordaApiVersion=5.3.0.8-alpha-1712154048069
```
Comment (filipesoliveira) [inline gradle.properties]: This change must be reverted before merging the PR!

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/services/internal/BackoffManagerImpl.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import net.corda.ledger.utxo.token.cache.services.BackoffManager
+import net.corda.utilities.retry.Exponential
+import net.corda.utilities.time.Clock
+import java.time.Duration
+import java.time.Instant
+import kotlin.math.min
+
+class BackoffManagerImpl(
+    private val clock: Clock,
+    private val dbBackoffMinPeriod: Duration,
```
Comment (vkolomeyko) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/services/internal/BackoffManagerImpl.kt]: Nit:
```suggestion
    private dbBackoffMinPeriod: Duration,
```

File: components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/services/internal/BackoffManagerImpl.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    private val dbBackoffMaxPeriod: Duration
+) : BackoffManager {
+
+    inner class AttemptAndBackoffTimePair(val attempt: Int, val backoffTime: Instant)
+
+    private val backoffStrategy = Exponential(base = 2.0, growthFactor = dbBackoffMinPeriod.toMillis())
```
Comment (vkolomeyko) [inline components/ledger/ledger-utxo-token-cache/src/main/kotlin/net/corda/ledger/utxo/token/cache/services/internal/BackoffManagerImpl.kt]: Nit:
```suggestion
    private val backoffStrategy = Exponential(growthFactor = dbBackoffMinPeriod.toMillis())
```

**Review by vkolomeyko — DISMISSED** [review]
LGTM - with few nitpicky changes suggested.

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Pull request merged**